include/s3select_json_parser.h

```diff
@@ -0,0 +1,475 @@
+    {} 
+
+    std::string get_key_path()
+    {//for debug
+	  std::string res;
+	  for(const auto & i: key_path)
+	  {
+	    res.append(i);
+	    res.append(std::string("/"));
+	  }
+	  return res;
+    }
+
+    void dec_key_path()
+    {
+      if (json_element_state.size())  {
+        if(json_element_state.back() != ARRAY_STATE)  {
+	        if(key_path.size() != 0) {
+	          key_path.pop_back();
+          }
+        }
+      }
+      
+      if (key_path.size() < from_clause.size()) {
+        prefix_match = false;
+      } 
+      else if (prefix_match) {
+          if (state == row_state::ARRAY || state == row_state::START_ROW) {
+            state = row_state::START_ROW;
+	    m_s3select_processing();
+            ++row_count;
+          }
+      }
+    }
+
+    void push_new_key_value(Valuesax& v)
+    { int json_idx =0; 
+
+      //std::cout << get_key_path() << std::endl;
+
+      auto predicate = [](std::string& it1, std::string& it2)
+			  {
+			    boost::to_upper(it1);
+			    boost::to_upper(it2);
+			    return it1 == it2;
+			  };
+
+      if (prefix_match) {
+        for (auto filter : query_matrix) {
+          if(std::equal(key_path.begin() + from_clause.size(), key_path.end(), filter.begin(),predicate)) {
+	    //TODO very intensove not need for key-path
+	    JsonParserHandler::json_key_value_t found_variable(key_path,v);
+            m_exact_match_cb(found_variable, json_idx);
+          }
+	  json_idx ++;//TODO can use filter - begin()
+        }
+      }
+      dec_key_path();
+    }
+
+    bool Null() {
+      // at this point should verify against from-clause/where-clause/project. if match then push to scratch-area
+      push_new_key_value(value.Parse(nullptr));
+      return true; }
+
+    bool Bool(bool b) {
+      push_new_key_value(value.Parse(b));
+      return true; }
+
+    bool Int(int i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint(unsigned u) {
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Int64(int64_t i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint64(uint64_t u) { 
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Double(double d) { 
+      push_new_key_value(value.Parse(d));
+      return true; }
+
+    bool String(const char* str, rapidjson::SizeType length, bool copy) {
+      //TODO use copy
+      push_new_key_value(value.Parse(str));
+      return true;
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool copy) {
+      key_path.push_back(std::string(str));
+      
+      if (key_path == from_clause || from_clause.size()==0) {
+        prefix_match = true;
+      }
+      return true;
+    }
+
+    bool StartObject() {      
+      if (key_path.size()) {
+        if (prefix_match && from_clause.size() && (key_path[key_path.size() -1] == from_clause[from_clause.size() - 1])) {
```

Comment:
need to test the following 
`from_clause` is zero ("")

`startObject` could only set `OBJECT_START_ROW`, if state in is not `START_ROW_*` state

Reply:
from-clause could be zero.

upon `StartObject` it starts a row only if the row is not already had started

Reply:
`(key_path[key_path.size() -1] == from_clause[from_clause.size() - 1])`
in the case `prefix_match` is true, the above expression is also true

Reply:
I am not sure of that. 
eg. from_clause: a/b
key_path: a/b/c
Here, prefix_match is true but, last_elements are different
This is helpful in 1 row definition where { }, and the row_count is 1